## Week 11, Lab 1: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the famous (but fake) Northwind database. It contains various tables of transactions. The schema for the Northwind database can be found [here](http://www.zentut.com/wp-content/uploads/downloads/2013/06/Northwind-Sample-Database-Diagram.pdf).

In [1]:
#!pip install psycopg3

In [2]:
import psycopg2

In [3]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine('postgres://jlpmxjkftcsepc:bb9cd75e6659d7d67ee3b86be0f905069319c75bd6823d786c808dd799ea85a2@ec2-54-227-251-33.compute-1.amazonaws.com:5432/d9teltkv30dni7')

### Prompt 1

How many rows are there in the `customers` table of the Northwind database?

In [4]:
# Use a SQL query to find the answer to the above prompt.
engine.dispose()

sql = """
SELECT COUNT(*)
FROM customers
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

count
0     91

**Prompt 1 Answer**: 91

### Prompt 2

If you sort the customers in alphabetical order by name, which customer is fourth?

In [5]:
# Use a SQL query to find the answer to the above prompt.
engine.dispose()

sql = """
SELECT contact_name
FROM customers c
ORDER BY c.contact_name ASC
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

contact_name
0          Alejandra Camino
1           Alexander Feuer
2         Anabela Domingues
3              Ana Trujillo
4             André Fonseca
5                 Ann Devon
6            Annette Roulet
7            Antonio Moreno
8                 Aria Cruz
9        Art Braunschweiger
10         Bernardo Batista
11           Carine Schmitt
12          Carlos González
13         Carlos Hernández
14          Catherine Dewey
15       Christina Berglund
16            Daniel Tonini
17               Diego Roel
18        Dominique Perrier
19         Eduardo Saavedra
20          Elizabeth Brown
21        Elizabeth Lincoln
22         Felipe Izquierdo
23          Francisco Chang
24              Fran Wilson
25       Frédérique Citeaux
26              Georg Pipps
27         Giovanni Rovelli
28      Guillermo Fernández
29               Hanna Moos
..                      ...
61          Matti Karttunen
62          Maurizio Moroni
63             Michael Holz
64     Miguel Angel Paolino
65              Palle Ibsen
66            Paolo Accorti
67         Pascale Cartrain
68         Patricia McKenna
69         Patricio Simpson
70            Paula Parente
71             Paula Wilson
72             Paul Henriot
73             Pedro Afonso
74            Peter Franken
75            Philip Cramer
76         Pirkko Koskitalo
77           Renate Messner
78            Rene Phillips
79              Rita Müller
80            Roland Mendel
81         Sergio Gutiérrez
82           Simon Crowther
83             Sven Ottlieb
84             Thomas Hardy
85        Victoria Ashworth
86                Yang Wang
87            Yoshi Latimer
88          Yoshi Tannamuri
89           Yvonne Moncada
90  Zbyszek Piestrzeniewicz

[91 rows x 1 columns]

**Prompt 2 Answer**: 3	Ana Trujillo

### Prompt 3

Among those customers who live in Germany, if you sort the customers in alphabetical order by name, which customer is fourth?

In [6]:
# Use a SQL query to find the answer to the above prompt.

engine.dispose()

sql = """
SELECT contact_name
FROM customers c
WHERE country ILIKE '%%Germany%%'
ORDER BY contact_name ASC
LIMIT 4
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

contact_name
0      Alexander Feuer
1           Hanna Moos
2  Henriette Pfalzheim
3          Horst Kloss

**Prompt 3 Answer**: Horst Kloss

### Prompt 4

From the `employees` table, what are the first names of the individuals who have a birthdate in the 1960s?

In [7]:
#Between 
# Use a SQL query to find the answer to the above prompt.
engine.dispose()

sql = """
SELECT first_name
FROM employees
WHERE birth_date BETWEEN '1-1-1960' AND '12-31-1969'
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

first_name
0      Janet
1    Michael
2     Robert
3       Anne

**Prompt 4 Answer**: Janet, Michael, Robert and Anne

### Prompt 5

In `order_details` table of the Northwind database, what is the total number of units sold for products 2, 3, and 4? 
> You should give us three numbers - one for each `ProductID`.

In [8]:
# Use a SQL query to find the answer to the above prompt.

engine.dispose()

sql = """
SELECT SUM(quantity)
FROM order_details
WHERE order_details.product_id = 2
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

engine.dispose()

sql = """
SELECT SUM(quantity)
FROM order_details
WHERE order_details.product_id = 3
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

engine.dispose()

sql = """
SELECT SUM(quantity)
FROM order_details
WHERE order_details.product_id = 4
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

sum
0  1057

sum
0  328

sum
0  453

**Prompt 5 Answer**: 1057, 328, 453

### Prompt 6

From the `order_details` table of the Northwind database, calculate the revenue per product. In your SQL output, rename the revenue column `Revenue` and sort your results from largest revenue per product to smallest revenue per product. Which `ProductID` has the fifth-highest revenue, and what is that revenue rounded to the nearest dollar? 
> Note that revenue for a given product should be equal to the total unit price of that product times the quantity. You do not need to worry about the discount here. If you haven't multiplied columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [9]:
# Use a SQL query to find the answer to the above prompt.
engine.dispose()


sql = """
SELECT order_details."product_id", SUM(order_details."quantity" * order_details."unit_price") AS "Revenue"
FROM order_details
GROUP BY order_details."product_id"
ORDER BY "Revenue" DESC
LIMIT 5
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

product_id        Revenue
0          38  149984.200821
1          29   87736.400513
2          59   76296.000000
3          60   50286.000374
4          62   49827.899998

**Prompt 6 Answer**: $49,828

### Prompt 7

From the `products` table of the Northwind database, show the `ProductName`, `UnitsInStock`, and `UnitsOnOrder` values for those products who have more units on order than units in stock. Among these results, which product has the most units on order?

In [10]:
sql = """
SELECT p.product_name, p.units_in_stock, p.units_on_order
FROM products as p
WHERE units_on_order > units_in_stock

"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

product_name  units_in_stock  units_on_order
0                       Chang              17              40
1               Aniseed Syrup              13              70
2              Queso Cabrales              22              30
3         Sir Rodney's Scones               3              40
4           Gorgonzola Telino               0              70
5          Mascarpone Fabioli               9              40
6                  Gravad lax              11              50
7                 Rogede sild               5              70
8                   Chocolade              15              70
9                    Maxilaku              10              60
10  Wimmers gute Semmelknödel              22              80
11  Louisiana Hot Spiced Okra               4             100
12        Scottish Longbreads               6              10
13              Longlife Tofu               4              20

**Prompt 7 Answer**: Louisiana Hot Spiced Okra

### Prompt 8

engine.dispose()

sql = """From the `order_details` table, show the products that have an average discount of at least 8%. Among these results, which average discount is closest to 8%?

In [12]:
engine.dispose()

sql = """

SELECT order_details."product_id", AVG(order_details."discount") 
FROM order_details
GROUP BY order_details."product_id"
HAVING AVG(order_details."discount") >= .08
ORDER BY AVG(order_details."discount") DESC
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()


product_id       avg
0          48  0.108333
1           2  0.102273
2           9  0.100000
3          58  0.100000
4          25  0.086111
5          30  0.081250

**Prompt 8 Answer**:  Among all Product IDs, Product ID 30 has the closest Discount to 8%, at .08125. 

### Prompt 9

Return the `orders` table, but also include the first name, last name, and country of the employee assigned to each order. What is the first and last name of the employee assigned to `OrderID` 11077? (This is the last row in the data, provided that you do not sort.)

In [ ]:
ORDER BY orderID DESC


In [16]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT orders.*, employees."first_name", employees."last_name", employees."country"
FROM orders
LEFT JOIN employees 
ON orders."employee_id" = employees."employee_id"


"""

pd.read_sql_query(sql, engine)

order_id customer_id  employee_id  order_date required_date shipped_date  \
0       10248       VINET            5  1996-07-04    1996-08-01   1996-07-16   
1       10249       TOMSP            6  1996-07-05    1996-08-16   1996-07-10   
2       10250       HANAR            4  1996-07-08    1996-08-05   1996-07-12   
3       10251       VICTE            3  1996-07-08    1996-08-05   1996-07-15   
4       10252       SUPRD            4  1996-07-09    1996-08-06   1996-07-11   
5       10253       HANAR            3  1996-07-10    1996-07-24   1996-07-16   
6       10254       CHOPS            5  1996-07-11    1996-08-08   1996-07-23   
7       10255       RICSU            9  1996-07-12    1996-08-09   1996-07-15   
8       10256       WELLI            3  1996-07-15    1996-08-12   1996-07-17   
9       10257       HILAA            4  1996-07-16    1996-08-13   1996-07-22   
10      10258       ERNSH            1  1996-07-17    1996-08-14   1996-07-23   
11      10259       CENTC            4  1996-07-18    1996-08-15   1996-07-25   
12      10260       OTTIK            4  1996-07-19    1996-08-16   1996-07-29   
13      10261       QUEDE            4  1996-07-19    1996-08-16   1996-07-30   
14      10262       RATTC            8  1996-07-22    1996-08-19   1996-07-25   
15      10263       ERNSH            9  1996-07-23    1996-08-20   1996-07-31   
16      10264       FOLKO            6  1996-07-24    1996-08-21   1996-08-23   
17      10265       BLONP            2  1996-07-25    1996-08-22   1996-08-12   
18      10266       WARTH            3  1996-07-26    1996-09-06   1996-07-31   
19      10267       FRANK            4  1996-07-29    1996-08-26   1996-08-06   
20      10268       GROSR            8  1996-07-30    1996-08-27   1996-08-02   
21      10269       WHITC            5  1996-07-31    1996-08-14   1996-08-09   
22      10270       WARTH            1  1996-08-01    1996-08-29   1996-08-02   
23      10271       SPLIR            6  1996-08-01    1996-08-29   1996-08-30   
24      10272       RATTC            6  1996-08-02    1996-08-30   1996-08-06   
25      10273       QUICK            3  1996-08-05    1996-09-02   1996-08-12   
26      10274       VINET            6  1996-08-06    1996-09-03   1996-08-16   
27      10275       MAGAA            1  1996-08-07    1996-09-04   1996-08-09   
28      10276       TORTU            8  1996-08-08    1996-08-22   1996-08-14   
29      10277       MORGK            2  1996-08-09    1996-09-06   1996-08-13   
..        ...         ...          ...         ...           ...          ...   
800     11048       BOTTM            7  1998-04-24    1998-05-22   1998-04-30   
801     11049       GOURL            3  1998-04-24    1998-05-22   1998-05-04   
802     11050       FOLKO            8  1998-04-27    1998-05-25   1998-05-05   
803     11051       LAMAI            7  1998-04-27    1998-05-25         None   
804     11052       HANAR            3  1998-04-27    1998-05-25   1998-05-01   
805     11053       PICCO            2  1998-04-27    1998-05-25   1998-04-29   
806     11054       CACTU            8  1998-04-28    1998-05-26         None   
807     11055       HILAA            7  1998-04-28    1998-05-26   1998-05-05   
808     11056       EASTC            8  1998-04-28    1998-05-12   1998-05-01   
809     11057       NORTS            3  1998-04-29    1998-05-27   1998-05-01   
810     11058       BLAUS            9  1998-04-29    1998-05-27         None   
811     11059       RICAR            2  1998-04-29    1998-06-10         None   
812     11060       FRANS            2  1998-04-30    1998-05-28   1998-05-04   
813     11061       GREAL            4  1998-04-30    1998-06-11         None   
814     11062       REGGC            4  1998-04-30    1998-05-28         None   
815     11063       HUNGO            3  1998-04-30    1998-05-28   1998-05-06   
816     11064       SAVEA            1  1998-05-01    1998-05-29   1998-05-04   
817     11065       LILAS           

**Prompt 9 Answer**: The Employee is Nancy	Davolio

### Prompt 10

Find the average amount of freight across all orders, broken out by country of *employee*. (**NOT `ShipCountry`**.) How much higher is the average freight from UK employees than the average freight from US employees?

> For a bonus, round the freight to the nearest whole number and rename the column `Average Freight`.

In [21]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT employees."country", AVG(orders."freight") AS "Average Freight"
FROM orders
LEFT JOIN employees
ON orders."employee_id" = employees."employee_id"
GROUP BY employees."country"

"""

pd.read_sql_query(sql, engine)

country  Average Freight
0      UK        78.977143
1     USA        77.973284

**Prompt 10 Answer**: